### PR TITLE
Sort out memory barriers for spin1api

### LINKED ISSUE #106: Gcc disable interrupt movement
Apparently, this bug (below) has been fixed. That should let us make the disable/re-enable interrupt code sequences be inline, which will help a lot in the sPyNNaker code. 

https://bugs.launchpad.net/gcc-arm-embedded/+bug/1722849

### PR DESCRIPTION
**Pull request by dkfellows**
Dear God, this is horribly subtle!

Fixes #106

### CODE REVIEW

**Comment by rowleya** [comment]
Testing with SpiNNakerManchester/sPyNNaker8#295

**Comment by rowleya** [comment]
I can't see any obvious issues with this, and the tests pass.  Is there any way that this could be tested by e.g. looking at the object code once compiled?

File: include/spin1_api.h [hunk 1]
```diff
@@ -385,20 +405,64 @@ __forceinline uint spin1_int_disable(void)
...
+    spin1_store_barrier();
+
+#ifdef __GNUC__
     asm volatile (
     "msr	cpsr_c, %[cpsr]"
     :
     : [cpsr] "r" (cpsr)
     :);
+#else
+    __asm { msr cpsr_c, sr }
```
Comment (rowleya) [inline include/spin1_api.h]:
```
armcc -c --c99 --cpu=5te --apcs interwork --min_array_alignment=4 -I /c/Users/zzalsar4/git/spinnaker_tools/include --split_sections -g -Ospace -o ../build/arm/spin1_isr.o spin1_isr.c
"c:/Users/zzalsar4/git/spinnaker_tools/include/spin1_api.h", line 422: Error:  #20: identifier "sr" is undefined
      __asm { msr cpsr_c, sr }
                          ^
```

Comment (rowleya) [reply include/spin1_api.h]: Ah, should be cpsr I assume

**Comment by rowleya** [comment]
I have done a quick test on this with gcc, and it seems to work.  Here is the code and output:

Without barrier:
```
#include <sark.h>
#include <spin1_api.h>

static int x;
static int y;

void c_main() {
    x = y + 1;
    y = 0;
}
```
Decompiled code - ```y = 0``` is moved before the addition:
```
 33c:	e59f3014 	ldr	r3, [pc, #20]	; 358 <RTR_DGC15+0x1c>
 340:	e3a01000 	mov	r1, #0
 344:	e5932004 	ldr	r2, [r3, #4]
 348:	e5831004 	str	r1, [r3, #4]
 34c:	e2822001 	add	r2, r2, #1
 350:	e5832000 	str	r2, [r3]
 354:	e12fff1e 	bx	lr
 358:	00400000 	.word	0x00400000
```

With barrier (note that spin1_store_barrier and spin1_memory_barrier are the same in GCC):
```
#include <sark.h>
#include <spin1_api.h>

static int x;
static int y;

void c_main() {
    x = y + 1;
    spin1_memory_barrier();
    y = 0;
}
```

Decompiled code - note ```y = 0``` is now after the addition:
```
 33c:	e59f3014 	ldr	r3, [pc, #20]	; 358 <RTR_DGC15+0x1c>
 340:	e5932004 	ldr	r2, [r3, #4]
 344:	e2822001 	add	r2, r2, #1
 348:	e5832000 	str	r2, [r3]
 34c:	e3a02000 	mov	r2, #0
 350:	e5832004 	str	r2, [r3, #4]
 354:	e12fff1e 	bx	lr
 358:	00400000 	.word	0x00400000
```

**Comment by rowleya** [comment]
A quick check with armcc - it doesn't reorder in the first place unfortunately so I can't confirm this works!

**Comment by lplana** [comment]
I'm very surprised that gcc moved the assignment to y!

The compiler should not be allowed to move an assignment to a variable around a use of that variable. It should flag that y may be used before a value is assigned, not move the assignment. This could lead to incorrect operation.

**Comment by rowleya** [comment]
Looking at the assembler, it doesn't really assign y to 0 first.  The difference is that it loads the value of 0 before the barrier in one example, and doesn't in the second.  So the operation is not compromised from what I can see.  The important thing here though is that the barrier does actually force the change from one to the other.

**Comment by lplana** [comment]
So the compiler is not really moving the assignment to y before the value is used in the addition. It's caching the value of y in r2 and using that value later on, which is correct operation. It is storing the new value of y before doing the addition, which is correct if there is no sequence point to enforce the ordering.

As you say, the memory barrier function seems to impose the correct ordering.

**Comment by rowleya** [comment]
The ordering is correct with gcc, but I still haven't got an example of it working for armcc, mostly because I think our version of the compiler is old enough that it doesn't bother with any fancy reordering with my simple examples!

File: include/spin1_api.h (hunk 1 above)
Comment (lplana) [reply include/spin1_api.h]: Version RVDS40 of armcc complains about the use of __cpsr__ in this instruction::

    "/spinnaker_tools/include/spin1_api.h", line 422: Error:  #1094: Expected a register expression
            __asm { msr cpsr_c, cpsr }
                                ^
    spin1_isr.c: 0 warnings, 1 error_

I suggest changing the parameter back to  __sr__  and updating the function prototype::

    SPIN1_INLINE void spin1_mode_restore(uint sr)

The GCC register map::

        : [cpsr] "r" (sr)

and the ARM instruction::

        __asm { msr cpsr_c, sr }

To get  rid of the compilation error.

**Comment by lplana** [comment]
I added a comment above about an armcc compilation error using version RVDS40 of the ARM tools:
```
ARM C/C++ Compiler, RVCT4.0 [Build 400]
```
As @rowleya compiled and tested it, I must be using a different, probably earlier, version of the compiler. I'm still suggesting that changes are made as, if I'm not mistaken, RVDS40 is the version of the tools available to the APT group.

Otherwise, I'm happy with the PR.

File: include/spin1_api.h (hunk 1 above)
Comment (rowleya) [reply include/spin1_api.h]: I think I have done this as requested:
https://github.com/SpiNNakerManchester/spinnaker_tools/pull/107/commits/6c6e6fdd6797d61d8e37de97154adc862a6997e7

Does this look correct?

Comment (rowleya) [reply include/spin1_api.h]: A merge with master changed the name to "value".  I am hoping this is now correct again

**Comment by rowleya** [comment]
Just merged in master, but it seems to have removed spin1_irq_enable.  Is this supposed to have happened?

File: include/spin1_api.h (hunk 1 above)
Comment (lplana) [reply include/spin1_api.h]: The code now compiles correctly with ARM C/C++ Compiler, RVCT4.0 [Build 400].

I did compile after the merge with master.

**Review by rowleya — APPROVED** [review]
I think this is ready now.  Tests are passing

**Comment by lplana** [comment]
> I think this is ready now. Tests are passing

Looks good to go to me too.

**Pull request merged**